src/Business Foundation/App/NoSeriesCopilot/src/Copilot/Tools/NoSeriesCopToolsImpl.Codeunit.al

```diff
@@ -0,0 +1,243 @@
+    end;
+
+    procedure SetFilterOnSetupTables(var TableMetadata: Record "Table Metadata")
+    begin
+        TableMetadata.SetFilter(Name, '* Setup');
+        TableMetadata.SetRange(ObsoleteState, TableMetadata.ObsoleteState::No); //TODO: Check if 'Pending' should be included
+        TableMetadata.SetRange(TableType, TableMetadata.TableType::Normal);
+    end;
+
+    procedure SetFilterOnNoSeriesFields(var TableMetadata: Record "Table Metadata"; var Field: Record "Field")
+    begin
+        Field.SetRange(TableNo, TableMetadata.ID);
+        Field.SetRange(Class, Field.Class::Normal);
+        Field.SetFilter(ObsoleteState, '<>%1', Field.ObsoleteState::Removed);
+        Field.SetRange(Type, Field.Type::Code);
+        Field.SetRange(Len, 20);
+        Field.SetRange(RelationTableNo, Database::"No. Series");
+    end;
+
+    procedure IsRelevant(TableMetadata: Record "Table Metadata"; Field: Record "Field"; Entities: List of [Text]): Boolean
+    var
+        Entity: Text;
+        String1: Text[250];
+        String2: Text[250];
+        Score: Decimal;
+        RecordMatchMgtCopy: Codeunit "Record Match Mgt. Copy"; //TODO: Replace with system app module when available
+    begin
+        //TODO: Replace this with embeddings, when Business Central supports it
+        foreach Entity in Entities do begin
+            String1 := RecordMatchMgtCopy.RemoveShortWords(RemoveTextPart(TableMetadata.Caption, ' Setup') + ' ' + RemoveTextParts(Field.FieldName, GetNoSeriesAbbreviations()));
+            String2 := RecordMatchMgtCopy.RemoveShortWords(Entity);
+            Score := RecordMatchMgtCopy.CalculateStringNearness(String1, String2, 1, 100) / 100;
+            if Score >= RequiredNearness() then
+                exit(true);
+        end;
+        exit(false);
+    end;
+
+    local procedure RequiredNearness(): Decimal
+    begin
+        exit(0.9)
+    end;
+
+    procedure AddChunkedTablesPrompt(var FinalPrompt: List of [Text]; var TablesPromptList: List of [Text]; var AddedCount: Integer)
+    var
+        TablePrompt: Text;
+        IncludedTablePrompts: List of [Text];
+    begin
+        // we add by chunks of 10 tables, not to exceed the token limit, as more than 10 tables can lead to hallucinations
+        foreach TablePrompt in TablesPromptList do
+            if (AddedCount < GetTablesChunkSize()) then begin
+                IncludedTablePrompts.Add(TablePrompt);
+                AddedCount += 1;
+            end;
+
+        foreach TablePrompt in IncludedTablePrompts do begin
+            FinalPrompt.Add(TablePrompt);
+            TablesPromptList.Remove(TablePrompt);
+        end;
+    end;
+
+    procedure RemoveTextParts(Text: Text; PartsToRemove: List of [Text]): Text
+    var
+        Part: Text;
+    begin
+        foreach Part in PartsToRemove do
+            Text := RemoveTextPart(Text, Part);
+        exit(Text);
+    end;
+
+    procedure RemoveTextPart(Text: Text; Part: Text): Text
```

Comment:
![image](https://github.com/microsoft/BCApps/assets/59221826/9e07c147-aebc-41a8-876c-19202ea33fd6)

So far only place where this case is existing is on Lookup triggers, everywhere else should be prefixed or affixed with something meaningful, Text: Text) : Text is just too confusing...